<?xml version="1.0"?>
<launch>
  <let name="project_name" value="simple_3d_localization"/>
  <let name="project_namespace" value="s3l"/>
  <let name="project_share" value="$(find-pkg-share $(var project_name))"/>

  <!-- global args -->
  <arg name="namespace" default=""/>
  <arg name="container_name" default="localization_container"/>
  <arg name="container_thread_num" default="2"/>
  <arg name="use_sim_time" default="false"/>

  <!-- launch args -->
  <!-- globalmap_server -->
  <arg name="globalmap_pcd" default="$(var project_share)/maps/GlobalMap.pcd"/>
  <arg name="convert_utm_to_local" default="false"/>
  <arg name="downsample_resolution" default="0.2"/>
  <!-- hdl_localization -->
  <arg name="use_intra_process_comms" default="false"/>
  <arg name="imu_topic" default="/livox/imu"/>
  <arg name="twist_topic" default="/odom/twist"/>
  <arg name="pointcloud_topic" default="velodyne_points"/>
  <arg name="global_map_topic" default="globalmap"/>
  <arg name="initialpose_topic" default="initialpose"/>
  <arg name="aligned_points_topic" default="aligned_points"/>
  <arg name="odom_topic" default="odom"/>
  <arg name="scan_matching_status_topic" default="scan_matching_status"/>

  <set_parameter name="use_sim_time" value="$(var use_sim_time)"/>

  <group>
    <node_container
      pkg="rclcpp_components"
      exec="component_container_mt"
      namespace="$(var namespace)"
      name="$(var container_name)"
      output="screen">
      <param name="thread_num" value="$(var container_thread_num)"/>
    </node_container>

    <!-- global_map -->
    <load_composable_node
      target="$(var namespace)/$(var container_name)">

      <composable_node
          pkg="$(var project_name)"
          plugin="$(var project_namespace)::map::GlobalmapServerNode"
          name="globalmap_server_node">
        <param name="globalmap_pcd" value="$(var globalmap_pcd)"/>
        <param name="convert_utm_to_local" value="$(var convert_utm_to_local)"/>
        <param name="downsample_resolution" value="$(var downsample_resolution)"/>
        <extra_arg name="use_intra_process_comms" value="$(var use_intra_process_comms)"/>
        <remap from="globalmap" to="$(var global_map_topic)"/>
      </composable_node>
      <composable_node
          pkg="$(var project_name)"
          plugin="$(var project_namespace)::hdl_localization::LocalizationNode"
          name="hdl_localization_node">
        <param from="$(var project_share)/config/hdl_localization.yaml" allow_substs="true"/>
        <remap from="imu/data" to="$(var imu_topic)"/>
        <remap from="odometry/twist" to="$(var twist_topic)"/>
        <remap from="points_raw" to="$(var pointcloud_topic)"/>
        <remap from="global_map" to="$(var global_map_topic)"/>
        <remap from="initialpose" to="$(var initialpose_topic)"/>
        <remap from="aligned_points" to="$(var aligned_points_topic)"/>
        <remap from="odom" to="$(var odom_topic)"/>
        <remap from="scan_matching_status" to="$(var scan_matching_status_topic)"/>
        <extra_arg name="use_intra_process_comms" value="$(var use_intra_process_comms)"/>
      </composable_node>
    </load_composable_node>
  </group>

</launch>